File: ios-native/DaidaiPanel/ui/settings/SponsorViewModel.swift
import Foundation

struct Sponsor: Identifiable {
    let id = UUID()
    let name: String
    let amount: String
    let message: String?
    let date: String?
}

@MainActor
final class SponsorViewModel: ObservableObject {
    @Published var sponsors: [Sponsor] = []
    @Published var totalAmount: String = ""
    @Published var isLoading = false
    @Published var error: String?

    private var api: ApiService

    init(api: ApiService) {
        self.api = api
    }

    func updateAPI(_ api: ApiService) {
        self.api = api
    }

    func load() async {
        isLoading = true
        error = nil
        do {
            let response: ApiResponse<[String: AnyCodable]> = try await api.sponsors()
            if let data = response.data {
                var list: [Sponsor] = []
                if let items = data["items"]?.value as? [[String: Any]] {
                    for item in items {
                        list.append(Sponsor(
                            name: (item["name"] as? String) ?? "匿名",
                            amount: (item["amount"] as? String) ?? "0",
                            message: item["message"] as? String,
                            date: item["date"] as? String
                        ))
                    }
                }
                self.sponsors = list
                self.totalAmount = (data["total"]?.value as? String) ?? "0"
            }
        } catch {
            self.error = ApiUtils.extractErrorMessage(from: error)
        }
        isLoading = false
    }
}

File: android-native/app/src/main/java/com/daidai/panel/ui/settings/SponsorViewModel.kt
package com.daidai.panel.ui.settings

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.daidai.panel.core.network.NetworkModule
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class Sponsor(
    val name: String = "",
    val amount: Double = 0.0,
    val message: String = "",
    val time: String = ""
)

data class SponsorState(
    val sponsors: List<Sponsor> = emptyList(),
    val total: Double = 0.0,
    val isLoading: Boolean = false,
    val error: String? = null
)

@HiltViewModel
class SponsorViewModel @Inject constructor(
    private val networkModule: NetworkModule
) : ViewModel() {

    private val _state = MutableStateFlow(SponsorState())
    val state: StateFlow<SponsorState> = _state.asStateFlow()

    init {
        load()
    }

    fun load() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, error = null)
            try {
                val api = networkModule.getApiService()
                val response = api.getSponsors()
                if (response.isSuccessful && response.body()?.isSuccess == true) {
                    @Suppress("UNCHECKED_CAST")
                    val rawList = response.body()?.data as? List<Map<String, Any>> ?: emptyList()
                    val sponsors = rawList.map { map ->
                        Sponsor(
                            name = map["name"] as? String ?: "",
                            amount = (map["amount"] as? Number)?.toDouble() ?: 0.0,
                            message = map["message"] as? String ?: "",
                            time = map["time"] as? String ?: ""
                        )
                    }
                    _state.value = _state.value.copy(
                        sponsors = sponsors,
                        total = sponsors.sumOf { it.amount },
                        isLoading = false
                    )
                } else {
                    _state.value = _state.value.copy(
                        isLoading = false,
                        error = response.body()?.message ?: "加载失败"
                    )
                }
            } catch (e: Exception) {
                _state.value = _state.value.copy(
                    isLoading = false,
                    error = e.message ?: "网络错误"
                )
            }
        }
    }

    fun clearError() {
        _state.value = _state.value.copy(error = null)
    }
}
